core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java

```diff
@@ -0,0 +1,181 @@
+package org.infinispan.expiration.impl;
+
+import static org.infinispan.test.TestingUtil.extractComponent;
+import static org.infinispan.test.TestingUtil.k;
+import static org.infinispan.test.TestingUtil.replaceComponent;
+import static org.infinispan.test.TestingUtil.v;
+import static org.testng.AssertJUnit.assertEquals;
+import static org.testng.AssertJUnit.assertFalse;
+import static org.testng.AssertJUnit.assertNull;
+import static org.testng.AssertJUnit.assertTrue;
+import static org.testng.AssertJUnit.fail;
+
+import java.lang.reflect.Method;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import org.infinispan.Cache;
+import org.infinispan.commons.time.TimeService;
+import org.infinispan.configuration.cache.CacheMode;
+import org.infinispan.configuration.cache.ConfigurationBuilder;
+import org.infinispan.container.entries.InternalCacheEntry;
+import org.infinispan.container.impl.InternalDataContainer;
+import org.infinispan.context.Flag;
+import org.infinispan.distribution.ch.KeyPartitioner;
+import org.infinispan.interceptors.locking.ClusteringDependentLogic;
+import org.infinispan.test.MultipleCacheManagersTest;
+import org.infinispan.transaction.LockingMode;
+import org.infinispan.util.ControlledTimeService;
+import org.testng.annotations.Test;
+
+/**
+ * Test to verify clustered max idle in a pessimistic transaction
+ *
+ * @author Pedro Ruivo
+ * @since 12.0
+ */
+@Test(groups = "functional", testName = "expiration.impl.MaxIdlePessimisticTxTest")
+public class MaxIdlePessimisticTxTest extends MultipleCacheManagersTest {
+
+   private static final int NUM_NODES = 3;
+   private static final long MAX_IDLE = 10;
+   private final ControlledTimeService timeService = new ControlledTimeService();
+
+   public void testWriteLock(Method method) throws Exception {
+      final String key = k(method);
+      final String value = v(method, 0);
+      final String value2 = v(method, 1);
+
+
+      Cache<String, String> cache = findNonOwnerCache(key);
+      cache.put(key, value, -1, TimeUnit.SECONDS, MAX_IDLE, TimeUnit.MILLISECONDS);
+
+      long lastWallClock = timeService.wallClockTime();
+      timeService.advance(MAX_IDLE - 1);
+
+      assertNotExpired(key);
+      assertLastUsed(key, lastWallClock);
+
+      cache.getAdvancedCache().getTransactionManager().begin();
+      assertEquals(value, cache.put(key, value2, -1, TimeUnit.SECONDS, MAX_IDLE, TimeUnit.MILLISECONDS));
+      assertEquals(value2, cache.get(key));
+      // rollback the transaction
+      cache.getAdvancedCache().getTransactionManager().rollback();
+
+      lastWallClock = timeService.wallClockTime();
+      assertNotExpired(key);
+      assertLastUsed(key, lastWallClock);
+
+      // if the tx didn't touch the key, this advance should expire it
```

Comment:
Actually, the comments are wrong (same comments, in all tests) :) my bad!

Not sure you notice, but there is a test missing: 
https://github.com/pruivo/infinispan/commit/1cdac799994e45fc5c01ae5ff6a5c3bf6346351e#diff-5932593a9461e5deec52370525a2cf35R144

Reply:
> Not sure you notice, but there is a test missing:
> [pruivo@1cdac79#diff-5932593a9461e5deec52370525a2cf35R144](https://github.com/pruivo/infinispan/commit/1cdac799994e45fc5c01ae5ff6a5c3bf6346351e#diff-5932593a9461e5deec52370525a2cf35R144)

Weird, I copied the file directly, must have had two commits before?

Reply:
```suggestion
      // the tx touches the key, this advance should not expire it
```